File: src/main/java/alg01unionfind/UF3WeightedQuickUnion.java
package alg01unionfind;

/**
 * <p>
 * To lessen the QuickUnion defect we should shorten the function <b>root</b> calls.
 * To do that we should avoid creating tall union trees, effectively reducing
 * the number of <b>while</b> loops.
 * </p>
 */
public class UF3WeightedQuickUnion extends UF2QuickUnion implements UnionFindAlgorithm {

    /**
     * To avoid creating tall trees we must never set a tall tree as a child of
     * a small tree growing the former even more. So we create treeSize array to keep
     * track of tree sizes
     */
    protected int[] treeSizeOf;

    public UF3WeightedQuickUnion(int size) {
        super(size);
        treeSizeOf = new int[size];
        for (int i = 0; i < size; i++) {
            // Initially all trees contain only one node
            treeSizeOf[i] = 1;
        }
    }

    /**
     * Now union and find will take at most O(log(N)) operations to complete,
     * as we reduced the size of trees.
     */
    @Override
    public void union(int n1, int n2) {
        int rootN1 = root(n1);
        int rootN2 = root(n2);
        // We put a smaller tree as a child to a taller tree
        if (treeSizeOf[rootN1] < treeSizeOf[rootN2]) {
            parentOf[rootN1] = rootN2;
            treeSizeOf[rootN2] += rootN1;
        } else {
            parentOf[rootN2] = rootN1;
            treeSizeOf[rootN1] += rootN2;
        }
    }

}

File: src/main/kotlin/alg01unionfind/UF3WeightedQuickUnion.kt
package alg01unionfind

/**
 * Addresses the main drawback of UF2QuickUnion: trees growing too big.
 * Avoid tree growth by storing tree sizes and checking it before attaching trees to each other.
 */
open class UF3WeightedQuickUnion(size: Int) : UF2QuickUnion(size) {

    // Initially all trees contain only one node
    private val sizeOfTree: Array<Int> = Array(size) { 1 }

    /**
     * Union and find will take at most O(log(N)) operations to complete,
     * as the size of the trees is reduced.
     */
    override fun union(n1: Int, n2: Int) {
        val n1TreeRoot = this.treeRoot(n1)
        val n2TreeRoot = this.treeRoot(n2)
        // A smaller tree is set as a child to a taller tree to avoid trees growth.
        if (this.sizeOfTree[n1TreeRoot] < this.sizeOfTree[n2TreeRoot]) {
            this.parentOf[n1TreeRoot] = n2TreeRoot
            this.sizeOfTree[n2TreeRoot] += this.sizeOfTree[n1TreeRoot]
        } else {
            this.parentOf[n2TreeRoot] = n1TreeRoot
            this.sizeOfTree[n1TreeRoot] += this.sizeOfTree[n2TreeRoot]
        }
    }
}
